src/common/utils/validators.py

```diff
@@ -35,125 +33,157 @@ def is_blueprint_instance_of(
     return False
 
 
-def _recursive_validate_single_attribute(
-    attribute: BlueprintAttribute, value: dict | bool | int | float | str, get_blueprint: Callable, key: str
-):
-    debug_message = f"Location: Entity in key '{key}'"
-    if attribute.is_primitive:
-        python_type = BuiltinDataTypes(attribute.attribute_type).to_py_type()
-        if not isinstance(value, python_type):
-            raise ValidationException(
-                f"Attribute '{attribute.name}' should be type '{python_type.__name__}'. Got '{type(value).__name__}'",
-                debug=debug_message,
-            )
-        if not attribute.is_optional:
-            if python_type == str and value == "":
-                raise ValidationException(f"Missing required attribute '{attribute.name}'", debug=debug_message)
+def _validate_primitive_attribute(attribute: BlueprintAttribute, value: bool | int | float | str, key: str):
+    python_type = BuiltinDataTypes(attribute.attribute_type).to_py_type()
+    if type(value) != python_type:
+        raise ValidationException(
+            f"Attribute '{attribute.name}' should be type '{python_type.__name__}'. Got '{type(value).__name__}'",
+            debug=_get_debug_message(key),
+        )
+    if not attribute.is_optional and python_type == str and value == "":
+        raise ValidationException(f"Missing required attribute '{attribute.name}'", debug=_get_debug_message(key))
 
-        return
-    if not attribute.is_optional and not value:
-        raise ValidationException(f"Missing required attribute '{attribute.name}'", debug=debug_message)
 
-    # Optional empty values are valid
-    if not value:
+def validate_entity_against_self(entity: dict | list, get_blueprint: Callable[..., Blueprint], key: str = "^") -> None:
+    if isinstance(entity, list):
+        # TODO: Check that dimensions is correct
+        for i, item in enumerate(entity):
+            validate_entity_against_self(item, get_blueprint, key=f"{key}.{i}")
         return
-
-    validate_entity(value, get_blueprint(attribute.attribute_type), get_blueprint, key)  # type: ignore
+    if not entity.get("type"):
+        raise ValidationException("Every entity must have a 'type' attribute", debug=_get_debug_message(key))
+    _validate_entity(entity, get_blueprint, get_blueprint(entity["type"]), "exact", key)
 
 
 def validate_entity(
     entity: dict | list,
-    blueprint: Blueprint,
     get_blueprint: Callable[..., Blueprint],
+    blueprint: Blueprint,
```

Comment:
Could get_blueprint be used to get correct blueprint instead of this extra blueprint argument?

Reply:
Then you would use validate_entity_against_self.